### DISCUSSION TITLE
I need to get the effective stress, it exists in my output, how do I call it in the source file and write it into my creep model

### DISCUSSION THREAD

**Discussion by Mrxiongjy** [idaholab/moose#27783] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I need to get the effective stress, it exists in my output, how do I call it in the source file and write it into my creep model

![0$C~0_E073{5DYQVXND4OBW](https://github.com/idaholab/moose/assets/113361478/4be274b7-af13-4196-ad18-821cab847323)
![8b6d3a5991eeac220eb9ee2a08cbb713](https://github.com/idaholab/moose/assets/113361478/5310277b-ac5b-4c60-bc10-c7fc01cab226)
![d3938487f678e412b4952a3fd21d6cba](https://github.com/idaholab/moose/assets/113361478/c7a154d1-fb8a-416b-bf9f-461fbc41ddaf)
![f4e1586b621b32ca2358a165dd4875f1](https://github.com/idaholab/moose/assets/113361478/28dc0085-d636-414c-b4c4-baa3f87614b6)

**Comment by GiudGiud** [comment]
Hello

It's defined as a material property already. 

You can see them all listed with Debug/show_material_props

**Reply by Mrxiongjy** [reply]
> Hello你好
> 
> It's defined as a material property already.它已经被定义为一种材料属性。
> 
> You can see them all listed with Debug/show_material_props您可以通过 Debug/show_material_props 看到它们全部列出

Do I need to enter debug mode? Sorry, I am not familiar with the use of this framework, can you give me a little more details?

**Reply by Mrxiongjy** [reply]
![1](https://github.com/idaholab/moose/assets/113361478/38b4930c-d08b-4504-ab9c-475f533c9b59)
![2](https://github.com/idaholab/moose/assets/113361478/ada16a2b-ab4a-45a3-a56c-633407487d4a)
I see them all listed with Debug/show_material_props ,then How do I call this vonmises_stress in another material and get its value

**Comment by Rahim-Habibi** [comment]
Hi, 
If you are writing source code for the creep model, you can simply define it in .h file and call it as a material property in .C file like: dumy(getMaterialProperty<RankTwoTensor>("dumy_name"))

**Reply by Mrxiongjy** [reply]
> Hi,你好， If you are writing source code for the creep model, you can simply define it in .h file and call it as a material property in .C file like: dumy(getMaterialProperty("dumy_name"))如果您正在为蠕变模型编写源代码，则可以简单地在 .h 文件中定义它，并将其作为 .C 文件中的材料属性调用，例如： dumy(getMaterialProperty("dumy_name"))

do you mean ,in .h file , const MaterialProperty<Real> & _vonmises_stress;  and then ,in .C file _vonmises_stress(getMaterialProperty<Real>("vonmises_stress"))
Is that okay?

**Reply by Rahim-Habibi** [reply]
const MaterialProperty <Real> & _vonmises_stress 
_vonmises_stress(getMaterialProperty<Real>("vonmises_stress"))

**Reply by GiudGiud** [reply]
This does look right

**Reply by Mrxiongjy** [reply]
> in .h file , const MaterialProperty & _vonmises_stress; and then ,in .C file _vonmises_stress(getMaterialProperty("vonmises_stress"))

const MaterialProperty<Real> & _vonmises_stress
_vonmises_stress(getMaterialProperty<Real>("vonmises_stress"))
when I do this ,then I get an error, As follows
![QQ图片20240606142433](https://github.com/idaholab/moose/assets/113361478/f8c26fde-a961-4fa8-83fa-f67d14cd53b6)

**Reply by GiudGiud** [reply, marked as answer]
I think you need to write:

this->template getMaterialProperty 

Because it's a template routine inside a templated class

**Reply by Mrxiongjy** [reply]
@GiudGiud @Rahim-Habibi thank you very much, I'm done.

**Reply by Mrxiongjy** [reply]
> I think you need to write:
> 
> this->template getMaterialProperty
> 
> Because it's a template routine inside a templated class

@GiudGiud Dear doctor , In the process of developing the creep strain model, I called the effective stress item and the compilation was successful, but I reported an error when running the input card, may I ask why
![Uploading QQ图片20240611191012.png…]()
![Uploading QQ图片20240611191021.png…]()

**Reply by GiudGiud** [reply]
hello

I dont see those pictures. I dont think they uploaded properly
If you are sharing a compilation error, it is much better to share these as copy pasted text, with triple back quotes before and after

Guillaume

**Reply by Mrxiongjy** [reply]
> hello
> 
> I dont see those pictures. I dont think they uploaded properly If you are sharing a compilation error, it is much better to share these as copy pasted text, with triple back quotes before and after
> 
> Guillaume

‘’‘ *** ERROR ***
Cyclic dependency detected in object ordering:
stress <- vonmises_stress_all <- eigen_strain_inner <- all_strain <- stress

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0 ’‘’

**Reply by Mrxiongjy** [reply]
and this is my .C file
![QQ图片20240611191012](https://github.com/idaholab/moose/assets/113361478/c57cbd64-7e51-4914-8838-5ce8462705b6)
![Uploading QQ图片20240611191021.png…]()

**Reply by GiudGiud** [reply]
The error comes from the chain of "declareMaterialProperty" and "getMaterialProperty" in the materials you use. 
It seems to be circular. Maybe it does not make sense to define the stress as depending on the vonmises stress?
You could:
- recompute the vonmises stress inside your stress material
- or chose to use getMaterialPropertyOld to lag one property by 1 timestep

**Reply by Mrxiongjy** [reply]
@GiudGiud Dear expert, Why is the contact pressure between pellets and cladding almost zero when the pellet and cladding are clearly deformed?
![Uploading b277b81fcb81eed11e69e91405046de3.JPG…]()

**Reply by GiudGiud** [reply]
didnt you make a separate post about this?

**Reply by Mrxiongjy** [reply]
yes，I do ，but there was no reply.（#27972）

**Reply by Mrxiongjy** [reply]
#27972

**Comment by wangzhaohao** [comment]
you can see my issue
https://github.com/idaholab/moose/discussions/27834#discussioncomment-9741246

**Reply by Mrxiongjy** [reply]
thank you

**Reply by Mrxiongjy** [reply]
When developing the gap thermal conductivity model, do you refer to GapConductanceConstraint or GapConductance? The former is used with the Mortar contact module

**Reply by wangzhaohao** [reply]
Previously, we used `GapConductance` now tring use former

**Reply by Mrxiongjy** [reply]
thank you again. I try to use mortar model for the contact of pellets and clad, Is this method feasible?
